rust-src/id/src/ffi.rs

```diff
@@ -292,6 +299,274 @@ pub extern "C" fn ip_info_cdi_verify_key(
+/// to valid utf8 strings.
+#[allow(clippy::too_many_arguments)]
+fn ar_info_create_helper(
+    identity: u32,
+    public_key_ptr: *mut u8,
+    public_key_len: size_t,
+    name_ptr: *const u8,
+    name_len: size_t,
+    url_ptr: *const u8,
+    url_len: size_t,
+    desc_ptr: *const u8,
+    desc_len: size_t,
+) -> Result<*mut ArInfo<G1>, anyhow::Error> {
+    // Identity.
+    let ar_identity = ArIdentity::try_from(identity)
+        .map_err(|_| anyhow::Error::msg("Unable to create ArIdentity instance from u32."))?;
+
+    // Public key.
+    let ar_public_key = {
+        let buf = &mut slice_from_c_bytes!(public_key_ptr, public_key_len as usize);
+
+        from_bytes::<elgamal::PublicKey<G1>, &[u8]>(buf)
+            .context("Unable to create PublicKey<G1> instance from byte array at public_key_ptr.")?
+    };
+
+    // Description.
+    let ar_description = {
+        let name = from_utf8(slice_from_c_bytes!(name_ptr, name_len as usize))
+            .context("Unable to decode byte array at name_ptr as utf8.")?
+            .to_string();
+
+        let url = from_utf8(slice_from_c_bytes!(url_ptr, url_len as usize))
+            .context("Unable to decode byte array at url_ptr as utf8.")?
+            .to_string();
+
+        let description = from_utf8(slice_from_c_bytes!(desc_ptr, desc_len as usize))
+            .context("Unable to decode byte array at desc_ptr as utf8.")?
+            .to_string();
+
+        Description {
+            name,
+            url,
+            description,
+        }
+    };
+
+    Ok(Box::into_raw(Box::new(ArInfo {
+        ar_identity,
+        ar_description,
+        ar_public_key,
+    })))
+}
+
+/// Attempt to create a [`GlobalContext`] instance from byte array pointers.
+/// Assumes that genesis_string_ptr points to a valid
+/// utf8 string, bulletproof_generators_ptr to a
+/// Generators<G1> instance and on_chain_commitments to a
+/// PedersenKey<G1> instance, and returns a null-pointer
+/// otherwise.
+#[no_mangle]
+unsafe extern "C" fn global_context_create(
+    genesis_string_ptr: *const u8,
+    genesis_string_len: size_t,
+    bulletproof_generators_ptr: *const u8,
+    bulletproof_generators_len: size_t,
+    on_chain_commitment_ptr: *const u8,
+    on_chain_commitment_len: size_t,
+) -> *mut GlobalContext<G1> {
+    if let Ok(ptr) = global_context_create_helper(
+        genesis_string_ptr,
+        genesis_string_len,
+        bulletproof_generators_ptr,
+        bulletproof_generators_len,
+        on_chain_commitment_ptr,
+        on_chain_commitment_len,
+    ) {
+        return ptr;
+    }
+    std::ptr::null_mut()
+}
+
+/// Attempt to create a [`GlobalContext`] instance from byte array pointers.
+/// Assumes that genesis_string_ptr points to a valid
+/// utf8 string, bulletproof_generators_ptr to a
+/// Generators<G1> instance and on_chain_commitments to a
+/// PedersenKey<G1> instance, and returns a null-pointer
+/// otherwise.
```

Comment:
This documentation does not reflect reality. Please update and explain that null is returned if data cannot be deserialized.